enhancements/installer/aws-custom-region-and-endpoints.md

```diff
@@ -0,0 +1,321 @@
+  // +optional
+  Ovirt *OvirtPlatformSpec `json:"ovirt,omitempty"`
+}
+
+
+// AWSPlatformSpec holds the current status of the Amazon Web Services infrastructure provider.
+type AWSPlatformSpec struct {
+  // APIEndpoints list contains custom endpoints which will override default
+  // service endpoint of AWS Services.
+  // There must be only one CustomEndpoint for a service.
+  // +optional
+  APIEndpoints []AWSCustomEndpoint `json:"apiEndpoints,omitempty"`
+}
+
+// AWSCustomEndpoint store the configuration of a custom url to
+// override existing defaults of AWS Services.
+// Currently Supports - EC2, IAM, ELB, S3 and Route53.
+type AWSCustomEndpoint struct {
+    Service string `json:"service"`
+
+    // This must be a HTTPS URL
+    URL     string `json:"url"`
+}
+```
+
+##### Drawbacks / Limitations / Future work
+
+1. Since there is no controller in the cluster that manages the state of `Infrastructure` object, we cannot ensure that the API endpoints in `spec` are synced to the `status` section, allowing cluster operators to continue to use the `status` section and provide a place to perform validations.
+
+2. There is no great way for the user to track if the change is propagated to the all the cluster operators.
+
+##### Alternatives
+
+1. Create another global configuration for Cloud API endpoints that stores information like the endpoints themselves, trusted bundles etc.
+Infrastructure global configuration already performs the function of tracking infrastructure related configuration and another global configuration that stores a part of the information doesn't seem like a great option. But it might allow validations and status observation by an independent controller.
+
+2. Configure each individual cluster operator
+There are five cluster operators that would need to be configured namely, cluster-kube-controller-manager, cluster-ingress-operator, cluster-machine-api-operator, cluster-image-registry-operator, cluster-credential-operator. There might be more operators like cluster-network-operator that might require access to the AWS APIs in the future to control the security group rules. Configuring all these separately is not a great UX for installer and a user who wants to modify the cluster to use API endpoints as day-2 operation.
+
+#### Cluster Kube Controller Manager Operator
+
+Currently the cluster-kube-controller-manager-operator uses the configmap `.spec.cloudConfig`, which stores the Kubernetes cloud-provider configuration, to configure the kube-controller-manager for various clouds.
+
+But since the API endpoints are now part of the `.spec.platformSpec.aws.apiEndpoints[]`, the kube-controller-manager will have to take the cloud configuration from the `.spec.cloudConfig` and add the [service overrides for AWS][k8s-aws-cloud-provider-service-overrides].
+
+If the `.spec.cloudConfig` already has the service overrides set, the operator should fail to roll out the configuration.
+
+#### Other Cluster Operators
```

Comment:
cc @dgoodwin @adambkaplan @ironcladlou @enxebre

Reply:
Adding a Jira for CCO.

Reply:
I assume that if an `S3` endpoint is not defined, the image-registry-operator will not be able to configure storage and will report itself `Degraded`.

Reply:
https://github.com/openshift/enhancements/pull/163/files/779c63898e5641a25ac914ac152d7fbdf088d242#diff-b6a7e34af40d81a977a31247b107348eR96-R99

these api endpoints are not required, but optional. so we continue to use the default ones unless these overrides are specified.